hpycc/connection.py

```diff
@@ -348,24 +362,60 @@ def get_logical_file_chunk(self, logical_file, start_row, n_rows,
             Maximum time, in seconds, to sleep between attempts.
             The true sleep time is a random int between 0 and
             `max_sleep`.
+        min_sleep: int
+            Maximum time, in seconds, to sleep between attempts.
+            The true sleep time is a random int between 'min_sleep' and
+            `max_sleep`.
+        temp_file: str, None
+            The file to write results to, see `low_mem` in `hpycc.get.get_thor_file`
+            for the logic behind this. If None then stores everything in memory.
+            Basically if present then we save RAM at the cost of speed.
+        cols: list
+            Column names for output. Required to ensure column order is maintained over
+            chunks.
 
         Returns
         -------
-        result_response: list of dicts
+        result_response: pd.DataFrame, str
             Rows of logical file as list of dicts. In the form
             [{"col1": 1, "col2": 2}, {"col1": 1, "col2": 2}, ...].
 
         """
+        # TODO lots of things about the docstring. DOn't make me look
+        # somewhere else for low mem. dont like that we need cols, dont like we
+        # need to specify temp file and min_sleep. should also just be an
+        # internal function
+
+        # TODO quote plus should be better named
+
         url = ("http://{}:{}/WsWorkunits/WUResult.json?LogicalName={}"
                "&Cluster=thor&Start={}&Count={}").format(
-            self.server, self.port, logical_file, start_row, n_rows)
-        r = self.run_url_request(url, max_attempts, max_sleep)
-        rj = r.json()
+            self.server, self.port, quote_plus(logical_file), start_row, n_rows)
+
+        resp = self.run_url_request(url, max_attempts, max_sleep, min_sleep)
+
+        # TODO do we need this, just let it fail
+        try:
+            resp = resp.json()
+        except JSONDecodeError:
+            print("Response unparsable: %s" % str(resp))
+
         try:
-            result_response = rj["WUResultResponse"]["Result"]["Row"]
-        except KeyError:
-            raise KeyError("json is : {}".format(rj))
-        return result_response
+            resp = resp["WUResultResponse"]["Result"]["Row"]
+        # TODO don't mask exceptions
+        except (KeyError, TypeError):
+            raise KeyError("json does not contain results of a WU:\n{}".format(resp))
+
+        resp = pd.DataFrame((item for item in resp))[cols]
+
+        # TODO I don;t like the temp_file, lock thing. can we use a temp folder somewhere else? it shouldnt be this functions job
+        if temp_file:
```

Comment:
Agree in principle but we need an alternative. Having it here gives us the best memory efficiency at the cost of speed. Any other solution I have involves keeping two copies of the data in RAM at some point which we should categorically avoid.

Reply:
so this function is mainly used in get_thor_file. If it returns a json we could write out a separate json per chunk using as_completed. this would only keep it in memory once. Then we could have a second step to read and collate these separate jsons. Does this sound possible, sensible, feasible? or just over complicated?

Reply:
I think I'll try some benchmarks on this, see what I can do how.

Reply:
I'm happy to discuss this in more detail too, I've only half thought through my solution